File: prototype_resource_game/src/systems/ResourceManager.ts
// src/systems/ResourceManager.ts
import { GAME_CONFIG } from '../config';

interface ResourceInventory {
    wood: number;
    stone: number;
    iron: number;
    food: number;
    water: number;
    circuit: number;
    aluminum: number;
    monsterParts: {
        fangs: number;
        hides: number;
        claws: number;
    };
}

export class ResourceManager {
    public resources: ResourceInventory;
    public dayResources: Partial<ResourceInventory>;
    public baseHP: number;
    public maxBaseHP: number;

    constructor() {
        this.resources = {
            wood: 0,
            stone: 0,
            iron: 0,
            food: GAME_CONFIG.STARTING_FOOD,
            water: 0,
            circuit: 0,
            aluminum: 0,
            monsterParts: {
                fangs: 0,
                hides: 0,
                claws: 0
            }
        };
        this.dayResources = {};
        this.baseHP = GAME_CONFIG.BASE_HP;
        this.maxBaseHP = GAME_CONFIG.BASE_HP;
    }

    addResource(type: string, amount: number): void {
        if (type === 'monsterParts' || type === 'monsterpart') return;
        
        const key = type as keyof Omit<ResourceInventory, 'monsterParts'>;
        if (key in this.resources) {
            this.resources[key] += amount;
            this.dayResources[key] = (this.dayResources[key] || 0) + amount;
        }
    }

    addMonsterPart(type: 'fangs' | 'hides' | 'claws', amount: number): void {
        this.resources.monsterParts[type] += amount;
    }

    getResource(type: string): number {
        const key = type as keyof Omit<ResourceInventory, 'monsterParts'>;
        if (key in this.resources) {
            return this.resources[key] || 0;
        }
        return 0;
    }

    loseDayResources(): Record<string, number> {
        const lost: Record<string, number> = {};
        for (const [type, amount] of Object.entries(this.dayResources)) {
            if (typeof amount === 'number' && amount > 0) {
                const loss = Math.floor(amount / 2);
                const key = type as keyof Omit<ResourceInventory, 'monsterParts'>;
                if (key in this.resources) {
                    this.resources[key] -= loss;
                    lost[type] = loss;
                }
            }
        }
        this.dayResources = {};
        return lost;
    }

    consumeFood(crewCount: number): { success: boolean; needed: number; shortage: number } {
        const needed = crewCount * 2;
        if (this.resources.food >= needed) {
            this.resources.food -= needed;
            return { success: true, needed, shortage: 0 };
        } else {
            const shortage = needed - this.resources.food;
            this.resources.food = 0;
            return { success: false, needed, shortage };
        }
    }

    takeBaseDamage(damage: number): boolean {
        this.baseHP -= damage;
        return this.baseHP <= 0;
    }

    resetDay(): void {
        this.dayResources = {};
    }
}

File: Unity-Projects/Assets/Day/Scripts/System/ResourceManager.cs
using System.Collections.Generic;
using UnityEngine;

public class ResourceManager : MonoBehaviour
{
    [System.NonSerialized]
    public Dictionary<string, int> Resources = new Dictionary<string, int>
    {
        ["wood"] = 0,
        ["food"] = 20,
        ["relic"] = 0
    };

    [System.NonSerialized]
    public Dictionary<string, int> DayResources = new Dictionary<string, int>();

    [System.NonSerialized]
    public Dictionary<string, int> MonsterParts = new Dictionary<string, int>
    {
        ["fangs"] = 0
    };

    public float BaseHp = 100f;
    public float MaxBaseHp = 100f;

    public void AddResource(string type, int amount)
    {
        if (!Resources.ContainsKey(type))
            return;

        Resources[type] += amount;

        if (DayResources.ContainsKey(type))
            DayResources[type] += amount;
        else
            DayResources[type] = amount;
    }

    public int GetResource(string type)
    {
        return Resources.ContainsKey(type) ? Resources[type] : 0;
    }

    public void AddMonsterPart(string type, int amount)
    {
        if (!MonsterParts.ContainsKey(type))
            return;

        MonsterParts[type] += amount;
    }

    public Dictionary<string, int> LoseDayResources()
    {
        var lost = new Dictionary<string, int>();

        foreach (var kvp in DayResources)
        {
            if (kvp.Value <= 0) continue;

            int loss = Mathf.FloorToInt(kvp.Value / 2f);
            if (loss <= 0) loss = 1;

            Resources[kvp.Key] -= loss;
            lost[kvp.Key] = loss;
        }

        DayResources.Clear();
        return lost;
    }

    /// <summary>
    /// Spends a resource outright (e.g. paying a recruit cost). Unlike
    /// AddResource with a negative amount, this does NOT touch DayResources —
    /// spending isn't "gained today" and shouldn't show up as a loss in the
    /// end-of-day summary or get halved by LoseDayResources(). Returns false
    /// (and spends nothing) if there isn't enough of the resource.
    /// </summary>
    public bool SpendResource(string type, int amount)
    {
        if (!Resources.ContainsKey(type)) return false;
        if (Resources[type] < amount) return false;

        Resources[type] -= amount;
        return true;
    }

    public bool ConsumeFood(int crewCount)
    {
        int needed = crewCount * 2;
        if (Resources["food"] >= needed)
        {
            Resources["food"] -= needed;
            return true;
        }

        Resources["food"] = 0;
        return false;
    }

    public bool TakeBaseDamage(float damage)
    {
        BaseHp -= damage;
        return BaseHp <= 0f;
    }
}